experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -0,0 +1,714 @@
+            if t == ZZ(1)
+                t = ZZ(2)
+            else
+                (_, q) = evenodd(q, t)
+            end
+        end
+        sqrt(BigInt(q)) > sqrt(sqrt(BigInt(N))) + 1 && return (true, (m, q))
+    end
+    return (false, (ZZ(1), N))
+end
+################################################################################
+
+function compute_ell_curve(N::fmpz, D::fmpz)
+    R = ResidueRing(ZZ, N)
+    if D == ZZ(-3)
+        return (R.([0, -1]), ZZ(1))
+    elseif D == ZZ(-4)
+        return (R.([-1, 0]), ZZ(1))
+    else
+        T, t = ZZ["t"]
+        p = hilbert_class_polynomial(Int(D), T)
+        p = change_base_ring(R, p)
+        j = roots(p)[1]
+        c = inv(j - R(1728)) * j
+        return ([-3c, -2c], ZZ(degree(p))) # Stores curve parameters and class number
+    end
+end
+
+################################################################################
+
+function random_point(v::Vector{fmpz_mod})
+    R = parent(v[1])
+    N = modulus(R)
+    A = rand(Int, 1000)
+    for a in A
+        z = R(a)^3 + R(a) * v[1] + v[2]
+        if jacobi_symbol(ZZ(z), N) != -1
+            (bool, res) = issquare_with_square_root(z)
+            if bool
+                return (bool, [R(a), res, R(1)])
+            end
+        end
+    end
+    return (false, R.([0, 0, 0]))
+end
+
+################################################################################
+
+function compute_p1p2(v::Vector{fmpz_mod}, m::fmpz, q::fmpz)
+    (bool, P) = random_point(v)
+    R = parent(v[1])
+    !bool && return (false, (R.([0, 0, 0]), ZZ(1)), (R.([0, 0, 0]), ZZ(1)))
+    P2 = _scalar_mult(P, v, div(m, q))
+    P1 = _scalar_mult(P2[1], v, div(m, q))
+    return (true, P1, P2)
+end
+
+################################################################################
+
+function quadratic_non_residue(N::fmpz, D::fmpz)
+    R = ResidueRing(ZZ, N)
+    A = rand(Int, 1000)
+    for a in A
+        (bool, res) = issquare_with_square_root(R(a))
+        if !bool
+            if D == ZZ(-3)
+                !isone(R(a)^div(N - 1, 3)) && return R(a)
+            else
+                return R(a)
+            end
+        end
+    end
+    return R(0)
+end
+
+################################################################################
+
+function atkin_morain_step(
+    N::fmpz,
+    D::fmpz,
+    m::fmpz,
+    q::fmpz,
+    res::Vector{Tuple{fmpz,Int}} = Vector{Tuple{fmpz,Int}}(),
+    pos::Int = 1,
+)
+    (v, h) = compute_ell_curve(N, D)
+    R = parent(v[1])
+    O_E = R.([0, 1, 0])
+    g = quadratic_non_residue(N, D)
+    k = 0
+
+    while k < h
+        P1 = (O_E, ZZ(1))
+        while P1[1] == O_E
+            (bool2, P1, P2) = compute_p1p2(v, m, q)
+            !bool2 && return (false, N, ZZ(1))
+            P2[1] == R.([0, 0, 0]) && return (false, N, P2[2])
+            P1[1] == R.([0, 0, 0]) && return (false, N, P1[2])
+            P2[1] != O_E && return (true, q, ZZ(1), push!(res, (q, pos)), pos)
+        end
+        k = k + 1
```

Comment:
I think `k` is only used to execute the loop (up to) `h` times, right? if so, the following is IMHO a bit clearer:
```suggestion

    for k in 1:h
        P1 = (O_E, ZZ(1))
        while P1[1] == O_E
            (bool2, P1, P2) = compute_p1p2(v, m, q)
            !bool2 && return (false, N, ZZ(1))
            P2[1] == R.([0, 0, 0]) && return (false, N, P2[2])
            P1[1] == R.([0, 0, 0]) && return (false, N, P1[2])
            P2[1] != O_E && return (true, q, ZZ(1), push!(res, (q, pos)), pos)
        end
```